ruby_event_store/lib/ruby_event_store/client.rb

```diff
@@ -174,31 +181,38 @@ def call
         unsubs += add_thread_subscribers
         @block.call
       ensure
-        unsubs.each(&:call)
+        unsubs.each(&:call) if unsubs
       end
 
       private
 
       def add_thread_subscribers
-        @subscribers.map do |handler, types|
-          @event_broker.add_thread_subscriber(handler, types)
+        @subscribers.map do |subscriber, types|
+          verify_subscriber(subscriber)
+          @event_broker.add_thread_subscriber(subscriber, types)
         end
       end
 
       def add_thread_global_subscribers
-        @global_subscribers.map do |s|
-          @event_broker.add_thread_global_subscriber(s)
+        @global_subscribers.map do |subscriber|
+          verify_subscriber(subscriber)
+          @event_broker.add_thread_global_subscriber(subscriber)
         end
       end
 
       def normalize_to_array(objs)
         return *objs
       end
+
+      def verify_subscriber(subscriber)
+        raise SubscriberNotExist if subscriber.nil?
+        @dispatcher.verify(subscriber)
+      end
     end
 
     def within(&block)
       raise ArgumentError if block.nil?
-      Within.new(block, event_broker)
+      Within.new(block, event_broker, dispatcher)
```

Comment:
step in a good direction, i wonder if we could not pass `self`.

Reply:
probably not.

Reply:
Nope. We do not have thread subscribers in client, we need a broker here (and dispatcher too). We might have get this from client but they are protected and exposing this outside client is not good.